src/tools/grpc_server.cc

```diff
@@ -165,26 +165,22 @@ class add_match {
     typedef std::set<std::string> line_set;
 
     add_match(CodeSearchResult* response)
-        : unique_lines_(NULL), response_(response) {}
-
-    add_match(line_set* unique_lines, CodeSearchResult* response)
-        : unique_lines_(unique_lines), response_(response) {}
+        : unique_lines_(new line_set), response_(response) {}
```

Comment:
This is a memory leak, since no one frees it. I think stack-declaring it in the parent and passing it in reference is probably the best option.

Reply:
What if `unique_lines_` were declared a `unique_ptr` instead of a bare old `*` — would that guarantee a free when the `add_match` object is removed from the stack?

Reply:
I think trying to use a `unique_ptr` in `add_match` is what led me to discover the terrible truth about `std::function` a long time ago: when an `add_match` is used as a `std::function`, a new `std::function` object is constructed, with a copy of the `add_match` object. But `unique_ptr` has no copy constructor, so using a `unique_ptr` as a field means that `add_match` also can't be copy-constructed, and so it fails at compilation time.

Reply:
Yep, @jboning is right. A `shared_ptr` would potentially work with approximately the desired semantics.